### PR TITLE
Added MMO-style protection areas.

### PR DESCRIPTION
**Pull request by biscuitWizard**
* Walls cannot be constructed/deconstructed in /area/protected where safe_zone is true
* Same for walls
* Same for anything made from a stack out of hand.
* Same for floors.
* Walls, floors, etc are all immune to damage. Machines cannot be deconstructed.

Anti-griefing tool. Could be used for timeout zones or something on a round-based server.

### CODE REVIEW

File: code/_helpers/protection.dm [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+		A = get_area(loc)
+		
+	if(!A.safe_zone)
+		return FALSE
+	if(!user)
+		return TRUE
+	// Check permission whitelists for this user.
+	return TRUE // TODO: actually implement.
+
+/area/protected
```
Comment (MistakeNot4892) [inline code/_helpers/protection.dm]: You don't typecheck `A`. This should really be an area flag rather than a bool.

**Comment by PsiOmegaDelta** [comment]
This implementation honestly feel very brittle, with checks needed everywhere.
That "everywhere"-part isn't a very good thing either.

**Comment by biscuitWizard** [comment]
> This implementation honestly feel very brittle, with checks needed everywhere.
> That "everywhere"-part isn't a very good thing either.

I don't feel this is a constructive comment.

File: code/_helpers/protection.dm (hunk 1 above)
Comment (biscuitWizard) [reply code/_helpers/protection.dm]: I wasn't aware of area_flags, I'll see about converting it, actually.

**Comment by PsiOmegaDelta** [comment]
Very well: There needs to be a more centralized way to check for protection than dropping proc-checks everywhere, be that via event registration or otherwise.

File: code/_helpers/protection.dm (hunk 1 above)
Comment (biscuitWizard) [reply code/_helpers/protection.dm]: Converted.

**Pull request closed without merging**